### ISSUE TITLE
fix(nargs): macro-obscured declarators report the macro's arity, not the function's

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#1213]
> **RESOLVED** — fixed on `fix/1213-macro-declarator-arity` (PR pending).
> Option **2** in its structural form, with the space keeping the macro's
> name (option **b**). See the
> [findings comment](https://github.com/dekobon/big-code-analysis/issues/1213#issuecomment-5196150672)
> for the corrections to the plan below: **there is no snapshot churn**
> (step 5 is wrong — all 46 affected spaces are under
> `DeepSpeech/tensorflow/**`, which the corpus test excludes), the gate
> must test **both** links' kinds and not just the inner one, 19 of the 46
> spaces are `ERROR`-recovery trees rather than macros, and the report's
> `operator()` count of 179 is actually 1,546.

## Summary

A macro-obscured declarator — the shape JNI shims use — parses as a
`function_declarator` nested inside another one:

```c
JNIEXPORT jlong RUN_STATS_METHOD(allocate)(JNIEnv *env, jclass clazz) { … }
```

```
function_declarator
├─ function_declarator          ← declarator field
│  ├─ identifier: RUN_STATS_METHOD
│  └─ parameter_list: (allocate)
└─ parameter_list: (JNIEnv *env, jclass clazz)
```

Since #1200, `nargs` reads the **innermost** declarator carrying a
`parameters` field, so this reports **1** argument (the macro's
`(allocate)`). Before #1200 the single-step read gave the outer list —
**2** here, 3 in the TensorFlow original. #1208 then made the *name*
come off that same innermost node, so the space is now called
`RUN_STATS_METHOD`.

The two surfaces agree, which is the property #1208 was after. The open
question is whether they agree on the right node.

## The two readings

**Syntactic (current).** The parser cannot know `RUN_STATS_METHOD` is a
macro. Read as written, `RUN_STATS_METHOD(allocate)` is the declarator
and `(JNIEnv *env, jclass clazz)` is the return type's parameter list.
One argument, named `RUN_STATS_METHOD`. Self-consistent, and identical
to how the same walk reads `int (*fp(int a, int b))(int c)`.

**Semantic.** After preprocessing this is one function taking the outer
list. Two arguments, and the name is unknowable without expanding the
macro. This is what a reader means by "how many arguments does that JNI
function take".

## Why it may matter

`bca check --threshold nargs=N` under-counts these functions, and every
JNI shim in a file collapses to one offender key
(`<anon@L28>::RUN_STATS_METHOD` × 4 in `run_stats_jni.cc`), where
overload-style key sharing is otherwise disambiguated by the baseline's
body hash (#377).

Measured over `tests/repositories/DeepSpeech` and `pdf.js`: 44 function
spaces take a name from a nested declarator, of which the macro spelling
is the large majority — `object_tracker_jni.cc` (18),
`imageutils_jni.cc` (5), `run_stats_jni.cc` (4).

## Options

1. **Leave it.** Syntactic is defensible, self-consistent, and cheap.
   Document the reading in the book's nargs page.
2. **Detect the shape** — a `function_declarator` whose `declarator`
   field is itself a `function_declarator` whose own parameter list
   holds a single bare `type_identifier` (no type, no declarator) is
   almost certainly a macro invocation, not a declarator. Prefer the
   outer list there. Narrow, but it is a heuristic keyed on the absence
   of a type, and `.claude/rules/grammar-dispatch.md` is unenthusiastic
   about those.
3. **Use the preprocessor data** `bca preproc` already collects for the
   C-family pipeline to decide whether the leading identifier is a known
   function-like macro. The most correct, and the most work.

No recommendation — this needs a call on how much the tool should model
the preprocessor.

## Provenance

Split out of #1208 rather than absorbed, per that issue's own scoping.

---

## Resolution Plan

The report asks for a call on how much the tool should model the
preprocessor. Four measurements below narrow that question
considerably: option 2 can be stated structurally rather than as a
heuristic, and option 3 is both more expensive and inert on the default
invocation.

### 1. The gate escape is real, and demonstrable

`run_stats_jni.cc` yields four spaces named `RUN_STATS_METHOD`
(L42, L49, L55, L72), each `nargs.function_args = 1`. Their true
arities are 2 (`allocate(env, clazz)`), 3 (`delete(env, clazz,
handle)`) and 4 (`add(env, clazz, handle, run_metadata)`).

```
$ bca check -p …/run_stats_jni.cc --no-config --threshold nargs=1
…:29-36: <anon@L28>::requireHandle: nargs = 2 (limit 1)
… 1 violation
```

Only the one ordinary function is flagged. All four JNI functions pass
a limit of 1 while declaring 2 to 4 arguments. At the shipped default
`nargs=5` nothing in this file changes either way, so the exposure is
confined to gates tightened to 3 or 4 — which is the range this repo's
own `#1143` discussion sits in.

### 2. Option 2 does not need to be a heuristic

The report proposes keying on "the inner parameter list holds a single
bare `type_identifier` (no type, no declarator)", and rightly notes
that `.claude/rules/grammar-dispatch.md` is unenthusiastic about
absence-keyed heuristics. That test is not required. The stronger gate:

> Prefer the outer parameter list whenever a `function_declarator`'s
> `declarator` field is **directly** another `function_declarator`.

That shape cannot arise from well-formed C or C++ — a function may not
return a function type (C11 6.7.6.3p1; C++ `[dcl.fct]`p11) — so every
legitimate function-returning-a-function-pointer form must interpose
parentheses. Confirmed by `bca dump` on the pinned grammars:

| source | chain |
| --- | --- |
| `int (*fp(int a, int b))(int c)` | `function_declarator → parenthesized_declarator → pointer_declarator → function_declarator` |
| `int (*g(void))[4]` | `array_declarator → parenthesized_declarator → pointer_declarator → function_declarator` |
| `void MACRO(name)(int a, int b)` | `function_declarator → function_declarator` **(direct)** |
| `void h(int a)(int b)` (ill-formed) | `function_declarator → function_declarator` **(direct)** |

A rule keyed on a shape the *language* forbids is a different thing
from one keyed on the absence of a type annotation. It also covers
more: a multi-argument macro spelling (`MACRO(a, b)(…)`) fails the
report's single-`type_identifier` test but is caught here.

### 3. False-positive sweep: one near-miss, and it is safe

Over `DeepSpeech` and `pdf.js`, the only construct that *looks* like
direct nesting in source text is C++ `operator()` — 179 occurrences.
It does not nest: the grammar emits a single `operator_name` node with
the parameter list as its sibling, not as a nested
`function_declarator`. Every genuine direct-nesting occurrence in the
corpora is a macro: `OBJECT_TRACKER_METHOD` (36), `IMAGEUTILS_METHOD`
(10), `RUN_STATS_METHOD` (5), `TFLITE_VARNAME_UNIQ`, `BM_NAME`,
`BM_LHS_NAME`, and two registration macros in headers. Zero non-macro
cases.

### 4. Option 3 is larger than it looks, and inert by default

Three findings, any one of which is disqualifying on its own:

- **The data is not there on the default path.** `PreprocResults` is
  populated only when the user passes `--preproc-data`
  (`load_preproc`, `big-code-analysis-cli/src/commands.rs:220`). A
  plain `bca metrics` or `bca check` passes `pr = None`, so a
  preprocessor-driven fix would not apply to the default invocation at
  all — only to the two-pass `bca preproc` → `--preproc-data`
  workflow.
- **The data does not carry what is needed.** `PreprocFile.macros` is
  a `HashSet<String>` of names (`src/preproc.rs:120`). It records
  neither function-like-ness nor arity.
- **The grammar cannot distinguish them either.**
  `tree-sitter-preproc/grammar.js:62` is
  `define: seq(/#define/, $.identifier, …, $.preproc_line)` — one rule,
  replacement text opaque. Teaching it the difference means a vendored
  grammar change with a regenerated `parser.c`, plus widening
  `PreprocFile`, which is public and serialized (`src/lib.rs:341`) —
  a public-API *and* on-disk-format break — plus wiring preproc data
  into the default run.

That is disproportionate to a shape §2 catches structurally.

### 5. The fix forces a decision #1208 deferred

#1208's stated invariant is that a function's name and its arity come
off the **same node**. Preferring the outer declarator for arity breaks
that: the outer node's `declarator` field is the inner
`function_declarator`, which is not an identifier kind, so every
getter's `matches!` gate rejects it and the space goes back to
nameless. Pick one:

- **(a) Arity outer, name `None`.** Honest — after `##` token pasting
  the real name (`Java_…_allocate`) is genuinely not in the source.
  Costs the 44 names #1208 just recovered.
- **(b) Arity outer, name still from the inner node.** Keeps
  `RUN_STATS_METHOD`, a stable greppable label, but retires "same
  node" as the invariant — it becomes "same function, two nodes".

**Recommend (b).** The name is what makes an offender row locatable,
and `RUN_STATS_METHOD` is the token a reader greps for. Note neither
choice fixes the offender-key collapse noted above: under (a) the four
rows collapse to `<anon@L28>` instead, which is worse.

### Plan (option 2, structural form)

1. Confirm the reading (semantic) and the name/arity pairing —
   (b) recommended. This is the human call the report asks for; steps
   2-8 assume it.
2. In `src/c_declarator.rs::innermost_declarator`, stop the chain when
   the current link's `declarator` field is itself a
   `function_declarator`. Compare by `kind()` string, not `kind_id`,
   per `.claude/rules/grammar-dispatch.md` §1 — `FunctionDeclarator2`
   / `3` aliases exist across the four C-family grammars and an id
   match would regress silently on the next bump. This is the same
   reason `CONVERSION_OPERATOR` is already a string constant.
3. Under (b), let `declarator_name` descend one link past where
   `innermost_declarator` now stops, so the name still resolves. Both
   the module doc and `declarator_name`'s doc assert the same-node
   pairing today and must be rewritten.
4. Tests, in both tables that cover this walk — `SHARED_SHAPES`
   (`src/c_declarator.rs`) and `c_declarator_shapes`
   (`src/metrics/nargs.rs:4555`):
   - `void MACRO(name)(int a, int b)` → nargs 2, name `MACRO`
     (row 3 of `SHARED_SHAPES` changes from the current syntactic
     expectation).
   - `void MACRO(a, b)(int x)` → the multi-argument macro the
     report's narrower gate would miss.
   - controls that must **not** move: `int (*fp(int a, int b))(int c)`
     → 1 arg / `fp`; `int (*g(void))[4]`; `operator()`.
   Test-via-revert per `.claude/rules/testing.md` — the controls are
   what prove the gate is not over-broad.
5. Integration snapshots: 44+ C-family spaces change arity across
   `DeepSpeech` and `pdf.js`, producing `.snap.new` **inside the
   `big-code-analysis-output` submodule**. Follow AGENTS.md's four-step
   discipline — accept, commit and push the submodule, record the new
   SHA in the *same* parent commit, and re-run after any rebase.
6. Docs: state the reading in the book's nargs section
   (`big-code-analysis-book/src/metrics.md:724`, and the gate
   subsection at `:749` which already explains what the threshold
   measures).
7. Refresh `.bca-baseline.toml` with
   `make self-scan-write-baseline-headroom` if any measured metric
   moves.
8. Sweep all four C-family languages. The walk is shared by C, C++,
   mozcpp and Objective-C, and the existing `SHARED_SHAPES` loop
   already runs each — confirm none is skipped by feature flags.

### If option 1 (leave it) is chosen

The work is step 6 alone: document the syntactic reading on the nargs
page and close. That remains defensible — it is self-consistent, it
matches how the same walk reads `int (*fp(int a, int b))(int c)`, and
it costs nothing.

## Assessment

| Dimension  | Rating |
|------------|--------|
| Difficulty | Medium |
| Complexity | Medium |
| Priority   | Medium |

Difficulty Medium: the walk change itself is a few lines, but it comes
with two test tables, a submodule snapshot bump, and doc rewrites of
the invariant #1208 just established. Complexity Medium rather than
High — the shared walk has two consumer surfaces and the snapshot churn
is real, but it is C-family only and no public API moves. Priority
Medium: the under-count is a genuine gate escape, but it is bounded to
C-family macro shims and invisible at the shipped `nargs=5` default,
and which reading is *correct* is the open question rather than a
settled bug.

**Comment by dekobon** [comment]
Added a resolution plan and assessment ratings (Difficulty Medium / Complexity Medium / Priority Medium). No `low-priority` label.

The issue asks for a call on how much the tool should model the preprocessor. Four measurements narrow that:

1. **The gate escape is demonstrable.** `bca check --threshold nargs=1` on `run_stats_jni.cc` flags only `requireHandle`; all four JNI functions report 1 while declaring 2, 3 and 4 arguments.
2. **Option 2 need not be a heuristic.** Rather than keying on the absence of a type in the inner list, key on the inner `function_declarator` being a *direct* child of the outer one. That shape cannot arise from well-formed C or C++ — a function may not return a function type (C11 6.7.6.3p1, C++ `[dcl.fct]`p11) — so every legitimate function-pointer return interposes a `parenthesized_declarator`. Verified by `bca dump` on the pinned grammars. It also catches multi-argument macro spellings the narrower test would miss.
3. **False-positive sweep is clean.** The only textual near-miss over DeepSpeech + pdf.js is C++ `operator()` (179 occurrences), which does not nest — the grammar emits a single `operator_name`. Every real direct-nesting case in the corpora is a macro.
4. **Option 3 is bigger than it looks and inert by default.** `PreprocResults` is populated only under `--preproc-data`, so a plain `bca metrics`/`check` run would not benefit at all; `PreprocFile.macros` records names only; and `tree-sitter-preproc`'s single `define` rule cannot tell function-like from object-like. The full path is a vendored grammar change plus a public, serialized API break.

One thing the plan surfaces that the report did not: the fix **breaks #1208's same-node name/arity invariant**. Preferring the outer declarator for arity leaves the name side on a non-identifier kind, so the space goes back to nameless unless the name is allowed to descend one link further. The plan recommends keeping the name (`RUN_STATS_METHOD`) and restating the invariant as same-*function*-two-nodes, but that is a human call and step 1 of the plan.

**Comment by dekobon** [comment]
## Fixed on `fix/1213-macro-declarator-arity`

The two calls this issue asked a human to make were taken as the plan
recommended: the **semantic** reading via option 2's structural form, and
**(b)** — arity from the outer declarator, name still the macro's.

### What the plan got right

Everything in §1–§4. Re-measured independently:

- `run_stats_jni.cc` reported `1, 1, 1, 1` against true arities `2, 3, 4,
  3`. `bca check --threshold nargs=1` found **1** violation there and now
  finds **5**.
- The direct nesting holds in all four grammars and is absent from
  `int (*fp(int a, int b))(int c)`, `int (*g(void))[4]` and `operator()`.
- Option 3 is disqualified three times over, exactly as described.

### What the plan got wrong

**Step 5 — there is no snapshot churn.** The plan says "44+ C-family
spaces change arity across `DeepSpeech` and `pdf.js`, producing
`.snap.new` **inside the submodule**", and prescribes the four-step
AGENTS.md discipline. None of that applies. A structural sweep of every
`function_definition` declarator chain in the corpora finds **46 affected
spaces in 15 files, all under `DeepSpeech/tensorflow/**`** — which
`tests/corpus/deepspeech_test.rs:25` excludes. `pdf.js` snapshots cover
`*.js` only and the repo holds **zero** C-family files outside DeepSpeech.
The corpus suite passes and the submodule is clean; no SHA bump.

**The report's `operator()` figure of 179 is wrong** by an order of
magnitude. `bca functions` reports **1,546** `operator()` spaces across
the corpora. The conclusion is unaffected and in fact stronger: not one
of the 46 direct nestings is an operator.

**Two further findings the plan did not have:**

1. **Both halves of the gate are load-bearing, and the plan only names
   one.** A pointer return puts the outer `function_declarator` in a
   `pointer_declarator`'s `declarator` field. A gate testing only "the
   inner link is a `function_declarator`" fires on the pointer, ends the
   chain on a node with no `parameters`, and regresses #1200's own
   `FILE *f(…)` and `int **g(…)` rows to **0**. The gate must test
   `current`'s kind too.

2. **19 of the 46 spaces are `ERROR`-recovery trees, not macros.**
   `TF_ASSIGN_OR_RETURN(bool ok, Try(x)); if (ok) { … }` recovers into
   the same direct nesting, with the `if` **condition** as the outer
   "parameter list". Those 19 move from the macro's argument count to the
   condition's — neither is an arity, and recovery has never been inside
   the walk's contract. There is no fixture: whether the grammar recovers
   this way depends on where the line breaks fall, since tree-sitter
   costs a recovery by the extent it skips, so any pinned spelling would
   be a claim about whitespace. Documented on the walk instead.

   The split is **27 macro shims / 19 recovery**, not the 26/20 an
   earlier pass of mine reported: `run_stats_jni.cc:49`
   (`RUN_STATS_METHOD(delete)` — `delete` is a keyword) carries its
   `ERROR` inside the *inner* list while its own chain is clean, so it is
   a shim the fix corrects.

### Implementation

`src/c_declarator.rs`, two functions, no getter and no public API change:

- `innermost_declarator` stops the chain when a `function_declarator`'s
  `declarator` field is directly another one. Compared by `kind()` string
  per `.claude/rules/grammar-dispatch.md` §1 — `tree-sitter-c` alone
  declares `FunctionDeclarator2` through `5`.
- `declarator_name` descends through a *run* of nested invocations, so
  `A(b)(c)(int x)` still resolves to `A`.

### Tests

Six new rows across `SHARED_SHAPES` and `c_declarator_shapes` /
`cpp_only_shapes`, so all four C-family languages run each. Six
perturbations, each against the full 3,325-test lib suite:

| perturbation | fails |
| --- | --- |
| gate arm deleted | the 4 macro `nargs` rows |
| `current.kind()` half dropped | every pointer-return row, incl. #1200's |
| `inner.kind()` half dropped | 8 rows in 2 tests |
| name descends 0 links | 3 macro name rows × 4 grammars |
| name descends 1 link, not a run | the `A(b, c)(d)(…)` row × 4 grammars |
| whole walk returns `None` | 28 rows |

That last one caught a row of my own that could not fail: an
`int (*g(void))[4]` control expecting `(0, 0)` — the value a dead walk
reports. Replaced with `int (*arr(int a, int b))[4]` → `(0, 2)`.

`BCA_GATE: pass`. Patch coverage 100% on added regions; the three
uncovered lines in `c_declarator.rs` are pre-existing test-failure paths.
